core/src/main/java/org/elasticsearch/common/network/NetworkService.java

```diff
@@ -106,31 +117,44 @@ public void addCustomNameResolver(CustomNameResolver customNameResolver) {
             }
         }
         // finally, fill with our default
-        if (bindHost == null) {
-            bindHost = DEFAULT_NETWORK_HOST;
+        if (bindHosts == null) {
+            bindHosts = new String[] { DEFAULT_NETWORK_HOST };
         }
-        InetAddress addresses[] = resolveInetAddress(bindHost);
+        InetAddress addresses[] = resolveInetAddresses(bindHosts);
 
         // try to deal with some (mis)configuration
-        if (addresses != null) {
-            for (InetAddress address : addresses) {
-                // check if its multicast: flat out mistake
-                if (address.isMulticastAddress()) {
-                    throw new IllegalArgumentException("bind address: {" + NetworkAddress.format(address) + "} is invalid: multicast address");
-                }
+        for (InetAddress address : addresses) {
+            // check if its multicast: flat out mistake
+            if (address.isMulticastAddress()) {
+                throw new IllegalArgumentException("bind address: {" + NetworkAddress.format(address) + "} is invalid: multicast address");
+            }
+            // check if its a wildcard address: this is only ok if its the only address!
+            if (address.isAnyLocalAddress() && addresses.length > 1) {
+                throw new IllegalArgumentException("bind address: {" + NetworkAddress.format(address) + "} is wildcard, but multiple addresses specified: this makes no sense");
             }
         }
         return addresses;
     }
 
+    /**
+     * Resolves {@code publishHosts} to a single publish address. The fact that it returns
+     * only one address is just a current limitation.
+     * <p>
+     * If {@code publishHosts} resolves to more than one address, <b>then one is selected with magic</b>,
+     * and the user is warned (they can always just be more specific).
+     * @param publishHosts list of hosts to publish as. this may contain special pseudo-hostnames
+     *                     such as _local_ (see the documentation). if it is null, it will be populated
+     *                     based on global default settings.
+     * @return single internet address
+     */
     // TODO: needs to be InetAddress[]
-    public InetAddress resolvePublishHostAddress(String publishHost) throws IOException {
+    public InetAddress resolvePublishHostAddresses(String publishHosts[]) throws IOException {
```

Comment:
It is only for now. I plan to solve the "multiple publish addresses" in a followup... but the api must be correct! 

This method takes as input, an array of addresses. and lots of places were passing `null` to it, so its important for type safety to rename!